server/webex_store.go

```diff
@@ -0,0 +1,257 @@
+
+	// Encrypt the refresh token if it's present
+	if session.Token.RefreshToken != "" {
+
+		refreshToken, err := encrypt([]byte(config.EncryptionKey), session.Token.RefreshToken)
+		if err != nil {
+			return err
+		}
+
+		session.Token.RefreshToken = refreshToken
+	}
+
+	data, err := json.Marshal(session)
+	if err != nil {
+		return err
+	}
+
+	if err := p.API.KVSet(fmt.Sprintf("%s%s", WebexOAuthSessionKey, session.UserID), data); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// loadWebexSession retrieves the webex session if present from the KV store
+func (p *Plugin) loadWebexSession(userID string) (*WebexOAuthSession, error) {
+	key := fmt.Sprintf("%s%s", WebexOAuthSessionKey, userID)
+
+	data, appErr := p.API.KVGet(key)
+	if appErr != nil {
+		p.API.LogError(
+			"Error retrieving webex session",
+			"user_id", userID,
+			"key", key,
+			"error", appErr.Error(),
+		)
+		return nil, appErr
+	}
+
+	if len(data) == 0 {
+		return nil, ErrWebexSessionNotFound
+	}
+
+	var session WebexOAuthSession
+
+	err := json.Unmarshal(data, &session)
+	if err != nil {
+		return nil, err
+	}
+
+	config := p.getConfiguration()
+
+	if time.Now().UTC().After(session.Token.Expiry) {
+		kvErr := p.API.KVDelete(key)
+		if kvErr != nil {
+			return nil, kvErr
+		}
+		return nil, ErrWebexSessionExpired
+	}
+
+	accessToken, err := decrypt([]byte(config.EncryptionKey), session.Token.AccessToken)
+	if err != nil {
+		return &session, err
+	}
+
+	session.Token.AccessToken = accessToken
+
+	// Decrypt refresh token if it's present
+	if session.Token.RefreshToken != "" {
+
+		refreshToken, err := decrypt([]byte(config.EncryptionKey), session.Token.RefreshToken)
+		if err != nil {
+			return &session, err
+		}
+
+		session.Token.RefreshToken = refreshToken
+	}
+
+	return &session, nil
+}
+
+// getWebexUserInfo retrieves the webex user info from the KV store and
+// if not present will request from the webex API and subsequently save
+// it to the KV store if successful
+func (p *Plugin) getWebexUserInfo(userID string) (*WebexUserInfo, error) {
+	webexUser, loadErr := p.loadWebexUser(userID)
+	if loadErr != nil {
+		if loadErr == ErrWebexUserNotFound {
+			session, err := p.loadWebexSession(userID)
+			if err != nil {
+				p.API.LogError("error looking up session", "error", err.Error())
+				return webexUser, err
+			}
+
+			webex, err := NewWebexClient(session.Token.AccessToken)
+			if err != nil {
+				p.API.LogError("Error creating new webex client", "error", err.Error())
+				return webexUser, err
+			}
+
+			person, _, err := webex.People.GetMe() // don't need resp, supressing it
+			if err != nil {
+				p.API.LogError("Error calling people.GetMe()", "error", err.Error())
+				return webexUser, err
+			}
+
+			webexUser = &WebexUserInfo{}
+			webexUser.FromWebexPerson(person)
+
+			err = p.storeWebexUser(userID, webexUser)
+			if err != nil {
+				p.API.LogError("Error saving webex user", "error", err.Error())
+				return webexUser, err
```

Comment:
here I'm not sure if we want to return `nil, err` or `webexUser, err` because in this case we can still use the webexUser. The question is, are we using it if we have an error?

Reply:
And the answer is "no", we only use the webexUser when `err == nil` so I would suggest or return here `nil, err` or handle this cases in the `handleOAuthCallback` handler. I think I prefer the former but I'm not 5/5 here.